atlasdb-config/src/main/java/com/palantir/atlasdb/factory/InitialisingTransactionManager.java

```diff
@@ -0,0 +1,667 @@
+            boolean allowHiddenTableAccess,
+            String userAgent) {
+
+        atlasFactory = new ServiceDiscoveringAtlasSupplier(config);
+        this.config = config;
+        this.runtimeConfigSupplier = runtimeConfigSupplier;
+        this.schemas = schemas;
+        this.env = env;
+        this.lockServerOptions = lockServerOptions;
+        this.allowHiddenTableAccess = allowHiddenTableAccess;
+        this.userAgent = userAgent;
+
+        LockRequest.setDefaultLockTimeout(
+                SimpleTimeDuration.of(config.getDefaultLockTimeoutSeconds(), TimeUnit.SECONDS));
+
+        registerEndpoints();
+
+        try {
+            initialise();
+        } catch (Throwable th) {
+            log.info("Synchronous initialisation failed, initialisation will be done asynchronously", th);
+            initialiseAsync();
+        }
+    }
+
+    private void registerEndpoints() {
+        env.register(remoteLockService);
+        env.register(timestampService);
+        if (config.leader().isPresent()) {
+            localPaxosServices = Leaders.createAndRegisterLocalServices(env, config.leader().get(), userAgent);
+        }
+        env.register(persistentLockService);
+        env.register(new CheckAndSetExceptionMapper());
+        env.register(sweeperService);
+    }
+
+    private void initialiseAsync() {
+        runAsync.accept(() -> {
+            while (uninitialised()) {
+                try {
+                    initialise();
+                } catch (Throwable th) {
+                    log.info("Async initialisation failed, retrying in {} seconds.", RETRY_AFTER_SECONDS, th);
+                    Uninterruptibles.sleepUninterruptibly(RETRY_AFTER_SECONDS, TimeUnit.SECONDS);
+                }
+            }
+        });
+    }
+
+    private void initialise() {
+        KeyValueService rawKvs = atlasFactory.getKeyValueService();
+
+        LockAndTimestampServices lockAndTimestampServices = createLockAndTimestampServices(
+                JavaSuppliers.compose(getTimestampClientConfigFunction(), runtimeConfigSupplier),
+                () -> LockServiceImpl.create(lockServerOptions),
+                atlasFactory::getTimestampService,
+                atlasFactory.getTimestampStoreInvalidator());
+
+        KeyValueService kvs = NamespacedKeyValueServices.wrapWithStaticNamespaceMappingKvs(rawKvs);
+        kvs = ProfilingKeyValueService.create(kvs, config.getKvsSlowLogThresholdMillis());
+        kvs = SweepStatsKeyValueService.create(kvs,
+                new TimelockTimestampServiceAdapter(lockAndTimestampServices.timelock()));
+        kvs = TracingKeyValueService.create(kvs);
+        kvs = AtlasDbMetrics.instrument(KeyValueService.class, kvs,
+                MetricRegistry.name(KeyValueService.class, userAgent));
+        kvs = ValidatingQueryRewritingKeyValueService.create(kvs);
+
+        TransactionTables.createTables(kvs);
+
+        persistentLockService.initialise(createPersistentLockService(kvs));
```

Comment:
I'm starting to wonder if we want to break out each component into an initializable object, such that we can skip initialization if it's already initialized. With the current logic, if something fails down the line (e.g., timelock), we then repeat all the previous initialization, even if it already succeeded. I think most of our stuff is idempotent, so not sure if that part is a primary concern, but currently we would leave the KVS and other `Closeable` things unclosed, which probably is bad.